### ISSUE TITLE
Fine tune first stage models on my own data set

### ISSUE THREAD

**Issue by ZhichaoTan** [nomoresomethingwentwrong/GLDM#2]
Hello, the GLDM project is indeed inspiring. However, when I attempt to fine-tune the first-stage models on my own dataset using settings ldm_con+wae_uncon.yml, I encounter the following issues:
```
       size mismatch for _full_graph_encoder._embed.weight: copying a param with shape torch.Size([166, 64]) from checkpoint, the shape in current model is torch.Size([147, 64]).
        size mismatch for _full_graph_encoder._model._first_layer.lins.0.weight: copying a param with shape torch.Size([64, 123]) from checkpoint, the shape in current model is torch.Size([64, 104]).
        size mismatch for _full_graph_encoder._model._first_layer.lins.1.weight: copying a param with shape torch.Size([64, 123]) from checkpoint, the shape in current model is torch.Size([64, 104]).
        size mismatch for _full_graph_encoder._model._first_layer.lins.2.weight: copying a param with shape torch.Size([64, 123]) from checkpoint, the shape in current model is torch.Size([64, 104]).
        size mismatch for _full_graph_encoder._model._first_layer.films.0.weight: copying a param with shape torch.Size([128, 123]) from checkpoint, the shape in current model is torch.Size([128, 104]).
        size mismatch for _full_graph_encoder._model._first_layer.films.1.weight: copying a param with shape torch.Size([128, 123]) from checkpoint, the shape in current model is torch.Size([128, 104]).
        size mismatch for _full_graph_encoder._model._first_layer.films.2.weight: copying a param with shape torch.Size([128, 123]) from checkpoint, the shape in current model is torch.Size([128, 104]).
        size mismatch for _full_graph_encoder._model._first_layer.lin_skip.weight: copying a param with shape torch.Size([64, 123]) from checkpoint, the shape in current model is torch.Size([64, 104]).
        size mismatch for _full_graph_encoder._model._first_layer.film_skip.weight: copying a param with shape torch.Size([128, 123]) from checkpoint, the shape in current model is torch.Size([128, 104]).
        size mismatch for _partial_graph_encoder._embed.weight: copying a param with shape torch.Size([166, 64]) from checkpoint, the shape in current model is torch.Size([147, 64]).
        size mismatch for _partial_graph_encoder._model._first_layer.lins.0.weight: copying a param with shape torch.Size([64, 124]) from checkpoint, the shape in current model is torch.Size([64, 105]).
        size mismatch for _partial_graph_encoder._model._first_layer.lins.1.weight: copying a param with shape torch.Size([64, 124]) from checkpoint, the shape in current model is torch.Size([64, 105]).
        size mismatch for _partial_graph_encoder._model._first_layer.lins.2.weight: copying a param with shape torch.Size([64, 124]) from checkpoint, the shape in current model is torch.Size([64, 105]).
        size mismatch for _partial_graph_encoder._model._first_layer.films.0.weight: copying a param with shape torch.Size([128, 124]) from checkpoint, the shape in current model is torch.Size([128, 105]).
        size mismatch for _partial_graph_encoder._model._first_layer.films.1.weight: copying a param with shape torch.Size([128, 124]) from checkpoint, the shape in current model is torch.Size([128, 105]).
        size mismatch for _partial_graph_encoder._model._first_layer.films.2.weight: copying a param with shape torch.Size([128, 124]) from checkpoint, the shape in current model is torch.Size([128, 105]).
        size mismatch for _partial_graph_encoder._model._first_layer.lin_skip.weight: copying a param with shape torch.Size([64, 124]) from checkpoint, the shape in current model is torch.Size([64, 105]).
        size mismatch for _partial_graph_encoder._model._first_layer.film_skip.weight: copying a param with shape torch.Size([128, 124]) from checkpoint, the shape in current model is torch.Size([128, 105]).
        size mismatch for _decoder._first_node_type_selector._final_layer.weight: copying a param with shape torch.Size([166, 256]) from checkpoint, the shape in current model is torch.Size([147, 256]).
        size mismatch for _decoder._first_node_type_selector._final_layer.bias: copying a param with shape torch.Size([166]) from checkpoint, the shape in current model is torch.Size([147]).
        size mismatch for _decoder._node_type_selector._final_layer.weight: copying a param with shape torch.Size([167, 256]) from checkpoint, the shape in current model is torch.Size([148, 256]).
        size mismatch for _decoder._node_type_selector._final_layer.bias: copying a param with shape torch.Size([167]) from checkpoint, the shape in current model is torch.Size([148]).
```
The _node_type_index_to_string in your project has a value of 166, whereas for my preprocessed data, it's 147. How can I modify it for my own fine-tuning?

**Comment by conghaowang** [comment]
Hello Zhichao, 

The length of the _node_type_index_to_string variable basically refers to the kinds of motifs/atoms in the dataset. If you want to use GLDM on your own dataset, I suggest you train the model from scratch since the type of motifs/atoms in your dataset differs from ours. 

Otherwise you can consider appending zeros to your node feature matrix along the feature dim to increase it from 147 to 166. But that means there are chances the model will predict the node type to be something that never exists in your dataset, so you will have to mask the appended features in the end. 

Hope this helps!

Cheers,
Conghao

**Comment by ZhichaoTan** [comment]
Thanks for your valuable advice.
But I still think the pretraining stage is of great importance due to data scarcity. Currently, I'm addressing this issue by preprocessing my own dataset following the MoLeR procedures. Notably, during preprocessing, I included the same metadata containing motifs/atoms information from GuacaMol to guarantee the same model architecture.
Another thing I want to ask is that the model in your article is also pretrained on GuccaMol and subsequently fine tuned on the L1000 dataset. In fact, these two datasets may also possess different motifs/atoms information, so does this effect the generation process and how do you deal with this issue?

**Comment by conghaowang** [comment]
I use the motifs/atoms extracted from the GuacaMol dataset when finetuing on the L1000 dataset. L1000 dataset is too small. It only contains around 7 thousand molecules, which covers a very limited chemical space for the model to learn. I think the motifs or atoms emerged in the L1000 dataset are also not representative enough. 

If your dataset is also small like L1000, it might be good to pretrain the model on a large dataset such as GuacaMol first and use the same motifs/atoms to finetune on the target dataset. 

Cheers, 
Conghao

**Comment by ZhichaoTan** [comment]
Thanks for your answer. I've learned a lot from your work.
